### PR TITLE
feat: search index tool

### PR DESCRIPTION
**Pull request by yuye-aws**
### Description

Implement SearchIndexTool to support agent framework. It takes in two string parameters: index and query. Here is the sample usage of this tool. Please remember to replace the model id and agent id.

```
POST /_plugins/_ml/agents/_register
{
  "name": "Test_Agent_For_ReAct",
  "type": "conversational",
  "description": "this is a test agent",
  "llm": {
    "model_id": "xxxxx",
    "parameters": {
      "max_iteration": 10,
      "stop_when_no_tool_found": true,
      "response_filter": "$.completion"
    }
  },
  "memory": {
    "type": "conversation_index"
  },
  "tools": [
    {
      "type": "SearchIndexTool",
      "description": "Use this tool to search index with a query. You should pass in two parameters: index and query. Index is the index name and the query is an OpenSearch DSL query."
    }
  ],
  "app_type": "my app"
}


POST /_plugins/_ml/agents/xxxxx/_execute
{
  "parameters": {
    "question": "Show me all the name field of all documents in index .plugins-ml-agent",
    "verbose": true
  }
}
```

The expected successful response should be 

```
{
  "inference_results": [
    {
      "output": [
        {
          "name": "memory_id",
          "result": "xxxxx"
        },
        {
          "name": "response",
          "result": "Thought: To get the name field of documents in the .plugins-ml-agent index, I should search that index using the SearchIndexTool."
        },
        {
          "name": "response",
          "result": "\nObservation: \"{\\\"_index\\\":\\\".plugins-ml-agent\\\",\\\"_source\\\":{\\\"name\\\":\\\"Test_Agent_For_ReAct\\\"},\\\"_id\\\":\\\"xxxxx\\\",\\\"_score\\\":1.0}\\n\""
        },
        {
          "name": "response",
          "result": "Based on the information I was previously provided from searching the index, the name field of the documents in the .plugins-ml-agent index contains 'Test_Agent_For_ReAct'."
        },
        {
          "name": "response",
          "result": "The name field of the documents in the .plugins-ml-agent index is 'Test_Agent_For_ReAct'."
        }
      ]
    }
  ]
}
```

If there is error when parsing the input, the expected error response could be:

```
{
  "error": {
    "root_cause": [
      {
        "type": "illegal_argument_exception",
        "reason": "[query] is null or empty, can not process it."
      }
    ],
    "type": "illegal_argument_exception",
    "reason": "[query] is null or empty, can not process it."
  },
  "status": 400
}
```

### Issues Resolved

1. Implement search index tool
2. Fix bugs in unit tests
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by zane-neo** [comment]
Please check the codecov report and add more UTs.

File: src/main/java/org/opensearch/agent/tools/SearchIndexTool.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+                if (hits != null && hits.length > 0) {
+                    StringBuilder contextBuilder = new StringBuilder();
+                    for (SearchHit hit : hits) {
+                        String doc = AccessController.doPrivileged((PrivilegedExceptionAction<String>) () -> {
+                            Map<String, Object> docContent = AbstractRetrieverTool.processResponse(hit);
+                            return StringUtils.gson.toJson(docContent);
+                        });
+                        contextBuilder.append(doc).append("\n");
+                    }
+                    listener.onResponse((T) StringUtils.gson.toJson(contextBuilder.toString()));
```
Comment (xinyual) [inline src/main/java/org/opensearch/agent/tools/SearchIndexTool.java:108]: Why we need a toJson here after calling toString?

Comment (yuye-aws) [reply src/main/java/org/opensearch/agent/tools/SearchIndexTool.java:108]: Sorry, there is no need for toJson. Thanks for pointing out.

**Comment by yuye-aws** [comment]
> Please check the codecov report and add more UTs.

I have add UT to cover more code. This PR is ready for review now.

File: src/main/java/org/opensearch/agent/tools/SearchIndexTool.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+@ToolAnnotation(SearchIndexTool.TYPE)
+public class SearchIndexTool implements Tool {
+
+    public static final String INPUT_FIELD = "input";
+    public static final String INDEX_FIELD = "index";
+    public static final String QUERY_FIELD = "query";
+
+    public static final String TYPE = "SearchIndexTool";
+    private static final String DEFAULT_DESCRIPTION =
+        "Use this tool to search index with a query. You should pass in two parameters: index and query. Index is the index name and the query is an OpenSearch DSL query.";
```
Comment (dhrubo-os) [inline src/main/java/org/opensearch/agent/tools/SearchIndexTool.java:50]: how about : `Use this tool to search an index by providing two parameters: 'index' for the index name, and 'query' for the OpenSearch DSL formatted query.`

File: src/main/java/org/opensearch/agent/tools/SearchIndexTool.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+            }, e -> {
+                log.error("Failed to search index", e);
+                listener.onFailure(e);
+            });
+
+            // since searching connector and model needs access control, we need
+            // to forward the request corresponding transport action
+            if (Objects.equals(index, ML_CONNECTOR_INDEX)) {
+                client.execute(MLConnectorSearchAction.INSTANCE, searchRequest, actionListener);
+            } else if (Objects.equals(index, ML_MODEL_INDEX)) {
```
Comment (zane-neo) [inline src/main/java/org/opensearch/agent/tools/SearchIndexTool.java:121]: Please add `ML_MODEL_GROUP_INDEX` as well, this one also needs permission control.

Comment (yuye-aws) [reply src/main/java/org/opensearch/agent/tools/SearchIndexTool.java:121]: `ML_MODEL_GROUP_INDEX ` will be processed with transport action `MLModelGroupSearchAction`

Comment (zane-neo) [reply src/main/java/org/opensearch/agent/tools/SearchIndexTool.java:121]: This is a tool and user could ask any questions including query model group which needs permission control, so you need to add model group index here just like the way handling model index by leveraging `MLModelGroupSearchAction.INSTANCE`.

**Review by zane-neo — APPROVED** [review]

**Pull request merged**